File: apps/ios/Loonext/Features/ForYou/LeadSourcesCard.swift
import SwiftUI

/**
 #301 — where these customers came from, on the home surface.

 Hand-port of `apps/web/src/components/for-you/lead-sources-card.tsx` and
 `LeadSourcesCard.kt`.

 # The coverage row is the design

 #301's fourth Acceptance line — "reporting distinguishes attributed from
 unknown, and never infers silently" — either happens on this panel or does
 not. A ranking built on a third of the conversations can be reordered
 completely by the other two thirds, and an owner acting on it spends real
 money on an artefact.

 So "Don't know" is a ROW, in the same list and on the same bar scale as the
 sources rather than a footnote under them. That is the only presentation in
 which an owner sees it competing with the channels they are about to spend on.

 # Absent rather than empty

 A quiet month renders NOTHING — not a zero, not an encouraging placeholder. A
 workspace that has named no sources gets one sentence about the cheapest way
 to start, because a table whose only row reads "Don't know: 40" is a scolding
 rather than a finding.
 */
struct LeadSourcesCard: View {
    let report: LeadSourceReport?
    /// #540: where "put a source on the numbers you advertise" happens.
    ///
    /// Not optional. A default of nil turns "nobody wired this" into a
    /// silently inert card rather than a compile error, and this is the state
    /// a NEW workspace opens on.
    let onSetUpSources: @MainActor () -> Void

    @Environment(\.appLocale) private var appLocale

    var body: some View {
        // Loading, or a month in which nothing happened. Silence, not a zero.
        if let report, report.total > 0 {
            // #540: the same heading treatment as the other three measures. Two of
            // the four carried their title inside the card and two above it, which
            // read as two different species of panel in one list.
            VStack(alignment: .leading, spacing: 0) {
            MeasureHeader(AppStrings.translate(appLocale, "inbox.leadSourcesTitle"))
            PaperCard {
                VStack(alignment: .leading, spacing: 0) {

                    // #232: unless the website is answering, in which case
                    // there IS something true to show and "you haven't set any
                    // sources up" would be a reproach aimed at somebody whose
                    // attribution is already working.
                    if report.sources.isEmpty && report.widget == 0 {
                        // Sources exist as a feature and this workspace has
                        // set none up, so every conversation is unknown.
                        Text(AppStrings.translate(appLocale, "inbox.leadSourcesNoneSetUp"))
                        .font(.footnote)
                        .foregroundStyle(.secondary)
                        .padding(.top, 6)
                        // #540: the paragraph tells a new crew what to do and
                        // web has always offered the door to do it. Both phones
                        // printed the instruction and stopped.
                        Button {
                            Haptics.tap()
                            onSetUpSources()
                        } label: {
                            HStack(spacing: 4) {
                                Text(
                                    AppStrings.translate(
                                        appLocale, "inbox.leadSourcesSetOneUp"
                                    )
                                )
                                Image(systemName: "arrow.right")
                                    .font(.scaled(12, weight: .semibold))
                            }
                            .font(.subheadline)
                            .foregroundStyle(BrandColor.olive)
                        }
                        .buttonStyle(.plain)
                        .padding(.top, 8)
                    } else {
                        body(for: report)
                    }
                }
                .padding(16)
            }
            }
            .padding(.top, 12)
        }
    }

    @ViewBuilder
    private func body(for report: LeadSourceReport) -> some View {
        let rows = leadSourceRows(report, locale: appLocale)
        let max = ([report.unknown, 1] + rows.map(\.total)).max() ?? 1

        if let headline = leadSourceHeadline(report, locale: appLocale) {
            Text(headline).font(.callout).padding(.top, 6)
        }
        if let note = report.note {
            Text(note).font(.footnote).padding(.top, 8)
        }

        VStack(spacing: 6) {
            ForEach(rows) { row in
                LeadSourceRow(name: row.name, total: row.total, max: max, muted: false)
            }
            if report.unknown > 0 {
                LeadSourceRow(
                    name: AppStrings.translate(appLocale, "inbox.leadSourcesUnknown"),
                    total: report.unknown,
                    max: max,
                    muted: true
                )
            }
        }
        .padding(.top, 10)

        // One sentence per number rather than an "s" glued on: French pluralises
        // the noun AND its article, which a suffix cannot reach.
        Text(
            AppStrings.translate(
                appLocale,
                report.total == 1
                    ? "inbox.leadSourcesFooterOne"
                    : "inbox.leadSourcesFooterMany",
                ["count": String(report.total)]
            )
        )
        .font(.footnote)
        .foregroundStyle(.secondary)
        .padding(.top, 10)
    }
}

private struct LeadSourceRow: View {
    let name: String
    let total: Int
    let max: Int
    let muted: Bool

    /**
     #238 — at a large Dynamic Type setting the BAR gives way, not the words.

     The row carries three things and they are not equal. The name says which
     channel and the number says how much; the bar restates the number as a
     length, for people who read a shape faster than a figure. When the reader's
     setting means all three cannot fit, keeping the decoration and truncating
     the label tells them a proportion about a source they can no longer
     identify — worse than no row at all.

     The Android twin was rendered at 200% and showed exactly that: "Your
     website" became "Your ..." and the count 14 broke across two lines as "1"
     over "4" — the number the row exists to convey, shown as a different
     number. These fixed 104 and 28 point frames are the same ones, so this
     carries the same fix rather than waiting for a device to prove it twice.

     *Applying: Zen of Clarity — when the space runs out, the decorative
     element goes first.*
     */
    @Environment(\.dynamicTypeSize) private var typeSize

    private var roomy: Bool { typeSize <= .large }

    var body: some View {
        HStack(spacing: 10) {
            Text(name)
                .font(.callout)
                .lineLimit(1)
                .truncationMode(.tail)
                .foregroundStyle(muted ? AnyShapeStyle(.secondary) : AnyShapeStyle(.primary))
                // The original 104 at ordinary sizes, untouched — it holds
                // "Word of mouth" and "Don't know" in full, and a proportional
                // width tried on Android was NARROWER and truncated both. A fix
                // for the largest setting must not make the common one worse.
                .frame(maxWidth: roomy ? 104 : .infinity, alignment: .leading)
            if roomy {
                GeometryReader { geo in
                    ZStack(alignment: .leading) {
                        Capsule()
                            .fill(Color.secondary.opacity(0.18))
                        Capsule()
                            .fill(muted ? Color.secondary.opacity(0.4) : BrandColor.olive.opacity(0.7))
                            .frame(
                                width: geo.size.width * CGFloat(total) / CGFloat(Swift.max(max, 1))
                            )
                    }
                }
                .frame(height: 6)
            }
            Text("\(total)")
                .font(.callout)
                .monospacedDigit()
                // A MINIMUM, not a fixed width: 28 points is narrower than two
                // digits at the largest settings, and a fixed frame folds them.
                .fixedSize()
                .frame(minWidth: 28, alignment: .trailing)
        }
    }
}

/** One row of the table: a source, or the folded tail. */
struct LeadSourceRowValue: Identifiable {
    let name: String
    let total: Int
    var id: String { name }
}

/** How many sources get their own row before the rest are folded together. */
private let leadSourceTopN = 4

/**
 The headline, in words, or nil when no honest one exists.

 Silent when the leading source is under a third of the attributed work: at
 that point "most of your work came from X" is simply false, and the table says
 it better than a wrong sentence would.
 */
/**
 Every channel this window can name, biggest first.

 #232 puts the website in HERE rather than pinning it under the configured
 sources. It is a channel like any other — a workspace whose site brings in most
 of the work should read that at the top of the list, and a row pinned last says
 the opposite by position while its number says otherwise. The server keeps it
 disjoint from `sources` precisely so it can be ranked against them without
 double-counting anybody.

 Same tie-break as the server's and as the other two clients', so equal counts
 do not shuffle between a phone and a laptop looking at the same month.

 `locale` is defaulted rather than required: the two callers below are a view
 that knows the locale and a test that does not care, and a new required
 parameter is how every `#Preview` in a file stops compiling.
 */
struct RankedChannel {
    /// The row label. A title, so it takes a capital.
    let name: String
    /// The same channel inside the headline sentence. "came from Your website"
    /// is the kind of thing only a rendered picture shows you; a source the
    /// workspace named themselves reads the same either way.
    let sentenceName: String
    let total: Int
}

func leadSourceChannels(
    _ report: LeadSourceReport,
    locale: String = "en"
) -> [RankedChannel] {
    var channels = report.sources.map {
        RankedChannel(name: $0.name, sentenceName: $0.name, total: $0.total)
    }
    if report.widget > 0 {
        channels.append(
            RankedChannel(
                name: AppStrings.translate(locale, "inbox.leadSourcesWebsite"),
                sentenceName: AppStrings.translate(locale, "inbox.leadSourcesWebsiteInline"),
                total: report.widget
            )
        )
    }
    return channels.sorted {
        $0.total != $1.total ? $0.total > $1.total : $0.name < $1.name
    }
}

func leadSourceHeadline(
    _ report: LeadSourceReport,
    locale: String = "en"
) -> String? {
    guard let top = leadSourceChannels(report, locale: locale).first else { return nil }
    let attributed = report.total - report.unknown
    guard attributed > 0 else { return nil }
    guard Double(top.total) / Double(attributed) >= 0.34 else { return nil }
    return "Most of the work you can account for came from \(top.sentenceName) — "
        + "\(top.total) of \(attributed)."
}

/** The rows to render: the top few, then everything else as one. */
func leadSourceRows(
    _ report: LeadSourceReport,
    locale: String = "en"
) -> [LeadSourceRowValue] {
    let ranked = leadSourceChannels(report, locale: locale)
    var rows = ranked.prefix(leadSourceTopN).map {
        LeadSourceRowValue(name: $0.name, total: $0.total)
    }
    let rest = ranked.dropFirst(leadSourceTopN)
    if !rest.isEmpty {
        rows.append(
            LeadSourceRowValue(
                name: "\(rest.count) more",
                total: rest.reduce(0) { $0 + $1.total }
            )
        )
    }
    return rows
}

File: apps/android/app/src/main/kotlin/com/loonext/android/features/foryou/LeadSourcesCard.kt
package com.loonext.android.features.foryou

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.outlined.ArrowForward
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.TextButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import com.loonext.android.core.i18n.AppStrings
import com.loonext.android.core.i18n.LocalAppLocale
import com.loonext.android.core.i18n.t
import com.loonext.android.core.model.LeadSourceReport
import com.loonext.android.ui.common.PaperCard
import com.loonext.android.ui.common.rememberHaptics
import com.loonext.android.ui.common.MeasureHeader

/**
 * #301 — where these customers came from, on the home surface.
 *
 * Hand-port of `apps/web/src/components/for-you/lead-sources-card.tsx`.
 *
 * # The coverage row is the design
 *
 * #301's fourth Acceptance line — "reporting distinguishes attributed from
 * unknown, and never infers silently" — either happens on this panel or does
 * not. A ranking built on a third of the conversations can be reordered
 * completely by the other two thirds, and an owner acting on it spends real
 * money on an artefact.
 *
 * So "Don't know" is a ROW, in the same list and on the same bar scale as the
 * sources rather than a footnote under them. That is the only presentation in
 * which an owner sees it competing with the channels they are about to spend
 * on.
 *
 * # Absent rather than empty
 *
 * A quiet month renders NOTHING — not a zero, not an encouraging placeholder.
 * A workspace that has named no sources gets one sentence about the cheapest
 * way to start, because a table whose only row reads "Don't know: 40" is a
 * scolding rather than a finding.
 */
@Composable
fun LeadSourcesCard(
    report: LeadSourceReport?,
    /**
     * #540: where "put a source on the numbers you advertise" actually
     * happens. REQUIRED rather than nullable-with-a-default, for the reason
     * `ForYouTab`'s own callbacks are (#503): a default of null turns "nobody
     * wired this" into a silently inert card instead of a compile error, and
     * this card is the one a NEW workspace opens on.
     */
    onSetUpSources: () -> Unit,
) {
    // Loading, or a month in which nothing happened. Silence, not a zero.
    if (report == null || report.total == 0) return

    // #540: the same heading treatment as the other three measures — see
    // PipelineCard for why the four have to agree.
    val locale = LocalAppLocale.current
    val haptics = rememberHaptics()

    Column(Modifier.fillMaxWidth().padding(top = 12.dp)) {
    MeasureHeader(t("inbox.leadSourcesTitle"))
    PaperCard(Modifier.fillMaxWidth()) {
        Column(Modifier.padding(16.dp)) {

            if (report.sources.isEmpty()) {
                // Sources exist as a feature and this workspace has set none
                // up, so every conversation is unknown.
                Text(
                    t("inbox.leadSourcesNoneSetUp"),
                    modifier = Modifier.padding(top = 6.dp),
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
                // #540: the paragraph tells a new crew what to do and web has
                // always offered the door to do it. Both phones printed the
                // instruction and stopped — the state a new workspace opens on,
                // explaining a task with no way to start it.
                TextButton(
                    onClick = {
                        haptics.tap()
                        onSetUpSources()
                    },
                    contentPadding = PaddingValues(horizontal = 0.dp, vertical = 4.dp),
                ) {
                    Text(t("inbox.leadSourcesSetOneUp"))
                    Icon(
                        Icons.AutoMirrored.Outlined.ArrowForward,
                        contentDescription = null,
                        modifier = Modifier.padding(start = 4.dp).size(14.dp),
                    )
                }
                return@Column
            }

            leadingSentence(report, locale)?.let { headline ->
                Text(
                    headline,
                    modifier = Modifier.padding(top = 6.dp),
                    style = MaterialTheme.typography.bodyMedium,
                )
            }

            report.note?.let { note ->
                Text(
                    note,
                    modifier = Modifier.padding(top = 8.dp),
                    style = MaterialTheme.typography.bodySmall,
                )
            }

            val rows = visibleRows(report, locale)
            val max = (rows.map { it.second } + report.unknown + 1).max()
            Spacer(Modifier.height(10.dp))
            rows.forEach { (name, total) ->
                SourceRow(name, total, max, muted = false)
            }
            if (report.unknown > 0) {
                SourceRow(t("inbox.leadSourcesUnknown"), report.unknown, max, muted = true)
            }

            Text(
                t(
                    if (report.total == 1) "inbox.leadSourcesFooterOne"
                    else "inbox.leadSourcesFooterMany",
                    "count" to report.total.toString(),
                ),
                modifier = Modifier.padding(top = 10.dp),
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }
    }
    }
}

@Composable
private fun SourceRow(name: String, total: Int, max: Int, muted: Boolean) {
    Row(
        Modifier.fillMaxWidth().padding(vertical = 3.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Text(
            name,
            modifier = Modifier.width(104.dp),
            maxLines = 1,
            overflow = TextOverflow.Ellipsis,
            style = MaterialTheme.typography.bodyMedium,
            color = if (muted) {
                MaterialTheme.colorScheme.onSurfaceVariant
            } else {
                MaterialTheme.colorScheme.onSurface
            },
        )
        Spacer(Modifier.width(10.dp))
        Box(
            Modifier
                .weight(1f)
                .height(6.dp)
                .clip(RoundedCornerShape(3.dp))
                .background(MaterialTheme.colorScheme.surfaceVariant),
        ) {
            Box(
                Modifier
                    .fillMaxWidth(total.toFloat() / max.toFloat())
                    .height(6.dp)
                    .clip(RoundedCornerShape(3.dp))
                    .background(
                        if (muted) {
                            MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = 0.4f)
                        } else {
                            MaterialTheme.colorScheme.primary.copy(alpha = 0.7f)
                        },
                    ),
            )
        }
        Spacer(Modifier.width(10.dp))
        Text(
            total.toString(),
            modifier = Modifier.width(28.dp),
            textAlign = TextAlign.End,
            style = MaterialTheme.typography.bodyMedium,
        )
    }
}

/** How many sources get their own row before the rest are folded together. */
private const val TOP_N = 4

/**
 * The headline, in words, or null when no honest one exists.
 *
 * Silent when the leading source is under a third of the attributed work: at
 * that point "most of your work came from X" is simply false, and the table
 * says it better than a wrong sentence would.
 */
internal fun leadingSentence(report: LeadSourceReport, locale: String): String? {
    val top = report.sources.firstOrNull() ?: return null
    val attributed = report.total - report.unknown
    if (attributed <= 0) return null
    if (top.total.toDouble() / attributed < 0.34) return null
    return AppStrings.translate(
        locale,
        "inbox.leadSourcesLeading",
        mapOf(
            "name" to top.name,
            "count" to top.total.toString(),
            "total" to attributed.toString(),
        ),
    )
}

/** The rows to render: the top few, then everything else as one. */
internal fun visibleRows(
    report: LeadSourceReport,
    locale: String,
): List<Pair<String, Int>> {
    val rows = report.sources.take(TOP_N).map { it.name to it.total }.toMutableList()
    val rest = report.sources.drop(TOP_N)
    if (rest.isNotEmpty()) {
        val label = AppStrings.translate(
            locale,
            "inbox.leadSourcesMore",
            mapOf("count" to rest.size.toString()),
        )
        rows.add(label to rest.sumOf { it.total })
    }
    return rows
}
